### PR TITLE
feat: parquet writer and schema evolution integration  test

### LINKED ISSUE #43: [Chore] Test cases for schema evolution & type change
We specifically need : 
1. Schema evolution (Adding new columns)
2. Data type changes of existing columns (Ex what if previous data type was string and now its a bool)? 
     - This can happen when alter column data type has happened in relational database.
     - Or Mongo has multiple document with same key but different types
3. Column name changes

### PR DESCRIPTION
**Pull request by vaibhav-datazip**
# Description

integration tests for parquet writer has been added from sources:
- postgres
- mysql
- mongodb

What this test case covers: 
- schema evolution for both parquet and iceberg
- parquet tests covering edge cases

Fixes #43 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] Tested in local test using all drivers individually 
- [x] Tested on GitHub ci running tests 
## Documentation

- [ ] Documentation Link: [link to README, olake.io/docs, or olake-docs]
- [x] N/A (bug fix, refactor, or test changes only)

### CODE REVIEW

File: utils/testutils/test_schema.go [hunk 1]
```diff
@@ -132,3 +132,137 @@ var GlobalTypeMapping = map[string]string{
 	"bit(n)":      "string",
 	"varying(n)":  "string",
 }
+
+// ParquetTypeMapping maps database types to Parquet types
+// Parquet has a simpler type system than Iceberg
+var ParquetTypeMapping = map[string]string{
```
Comment (vaibhav-datazip) [inline utils/testutils/test_schema.go]: can be removed if same for both iceberg and parquet

File: drivers/postgres/internal/postgres_test_util.go [hunk 2]
```diff
@@ -256,7 +288,34 @@ var ExpectedUpdatedPostgresData = map[string]interface{}{
 	"col_xml":               "<updated>value</updated>",
 }
 
-var PostgresToIcebergSchema = map[string]string{
+var ExpectedParquetUpdateData = map[string]interface{}{
```
Comment (vaibhav-datazip) [inline drivers/postgres/internal/postgres_test_util.go]:
```suggestion
var ExpectedParquetUpdatedData = map[string]interface{}{
```

File: drivers/mongodb/internal/mon_test.go [hunk 3]
```diff
@@ -10,13 +10,14 @@ import (
...
-		ExpectedUpdateData: ExpectedUpdatedMongoData,
-		DataTypeSchema:     MongoToIcebergSchema,
-		ExecuteQuery:       ExecuteQuery,
-		IcebergDB:          "mongodb_olake_mongodb_test",
+		TestConfig:                testutils.GetTestConfig(string(constants.MongoDB)),
+		Namespace:                 "olake_mongodb_test",
+		ExpectedData:              ExpectedMongoData,
+		ExpectedIcebergUpdateData: ExpectedIcebergUpdatedData,
+		ExpectedParquetUpdateData: ExpectedParquetUpdatedData,
+		DestinationDataTypeSchema: MongoToDestinationSchema, // Use same schema for parquet
```
Comment (vaibhav-datazip) [inline drivers/mongodb/internal/mon_test.go]: can remove unnecessary comment

File: drivers/mongodb/internal/testdata/test_streams.json [hunk 4]
```diff
@@ -1 +1 @@
...
...tamp"},"name_varchar":{"type":["string"],"destination_column_name":"name_varchar"}}},"supported_sync_modes":["cdc","strict_cdc","full_refresh","incremental"],"source_defined_primary_key":["_id"],"available_cursor_fields":["id_bigint","created_timestamp","id_maxkey","id_double","id_nested","id","id_nil","id_timestamp","id_int","id_bool","name_varchar","id_minkey","id_regex","_id"],"sync_mode":"cdc","destination_database":"mongodb:olake_mongodb_test","destination_table":"mongodb_test_table_olake"}},{"stream":{"name":"test_collection","namespace":"olake_mongodb_test","type_schema":{"properties":{"_cdc_timestamp":{"type":["timestamp_micro","null"],"destination_column_name":"_cdc_timestamp"},"_olake_id":{"type":["string","null"],"destination_column_name":"_olake_id"},"_olake_timestamp":{"type":["null","timestamp_micro"],"destination_column_name":"_olake_timestamp"},"_op_type":{"type":["string","null"],"destination_column_name":"_op_type"}}},"supported_sync_modes":["full_refresh","incremental","cdc","strict_cdc"],"source_defined_primary_key":["_id"],"available_cursor_fields":[],"sync_mode":"cdc","destination_database":"mongodb:olake_mongodb_test","destination_table":"test_collection"}}]}
```
Comment (vikaxsh) [inline drivers/mongodb/internal/testdata/test_streams.json:1]: what is need of "id"?

File: destination/parquet/parquet.go [hunk 5]
```diff
@@ -186,8 +186,11 @@ func (p *Parquet) Write(_ context.Context, records []types.RawRecord) error {
 
 // Check validates local paths and S3 credentials if applicable.
 func (p *Parquet) Check(_ context.Context) error {
+	uniqueSuffix := fmt.Sprintf("%d", time.Now().UnixNano())
+	threadID := fmt.Sprintf("test_parquet_destination_%s", uniqueSuffix)
+
 	p.options = &destination.Options{
-		ThreadID: "test_parquet_destination",
+		ThreadID: threadID,
```
Comment (vikaxsh) [inline destination/parquet/parquet.go]: why?

File: drivers/mongodb/internal/mon_test_util.go [hunk 6]
```diff
@@ -194,10 +233,8 @@ func insertTestData(t *testing.T, ctx context.Context, collection *mongo.Collect
 			"id_minkey":         primitive.MinKey{},
 			"id_maxkey":         primitive.MaxKey{},
 			"name_varchar":      "varchar_val",
-		},
-	}
+		}
```
Comment (vikaxsh) [inline drivers/mongodb/internal/mon_test_util.go:199]: why?

File: drivers/mongodb/internal/mon_test_util.go [hunk 7]
```diff
@@ -217,9 +254,9 @@ var ExpectedMongoData = map[string]interface{}{
 	"name_varchar":      "varchar_val",
 }
 
-var ExpectedUpdatedMongoData = map[string]interface{}{
+var ExpectedIcebergUpdatedData = map[string]interface{}{
 	"id_bigint":         int64(987654321098765),
-	"id_int":            int32(200),
+	"id_int":            int64(200),
```
Comment (vikaxsh) [inline drivers/mongodb/internal/mon_test_util.go]: how we would test for int32?

File: drivers/mongodb/internal/mon_test.go [hunk 8]
```diff
@@ -10,25 +10,26 @@ import (
...
+// 		Namespace:       "twitter_data",
+// 		BackfillStreams: []string{"tweets"},
+// 		CDCStreams:      []string{"tweets_cdc"},
+// 		ExecuteQuery:    ExecuteQuery,
+// 	}
 
-	config.TestPerformance(t)
-}
+// 	config.TestPerformance(t)
+// }
```
Comment (vikaxsh) [inline drivers/mongodb/internal/mon_test.go]: why commented?

File: drivers/mongodb/internal/mon_test_util.go [hunk 9]
```diff
@@ -105,11 +105,11 @@ func ExecuteQuery(ctx context.Context, t *testing.T, streams []string, operation
 		require.NoError(t, err, "Failed to insert document")
 
 	case "update":
-		filter := bson.M{"id_int": int32(100)}
+		filter := bson.M{"id": int32(1)}
 		update := bson.M{
 			"$set": bson.M{
 				"id_bigint":         int64(987654321098765),
-				"id_int":            int32(200),
+				"id_int":            int64(200),
```
Comment (vikaxsh) [inline drivers/mongodb/internal/mon_test_util.go:112]: we should keep different test case of evolve test and record update

File: drivers/mysql/internal/mysql_test_util.go [hunk 10]
```diff
@@ -130,10 +130,10 @@ func ExecuteQuery(ctx context.Context, t *testing.T, streams []string, operation
 				created_date = '2024-07-01 15:30:00',
 				created_timestamp = '2024-07-01 15:30:00', is_active = 0,
 				long_varchar = 'updated long...', name_bool = 0
-			WHERE id = 1`, integrationTestTable)
+			WHERE id = 6`, integrationTestTable)
 
 	case "delete":
-		query = fmt.Sprintf("DELETE FROM %s WHERE id = 1", integrationTestTable)
+		query = fmt.Sprintf("DELETE FROM %s WHERE id = 6", integrationTestTable)
```
Comment (vikaxsh) [inline drivers/mysql/internal/mysql_test_util.go]: why?

File: utils/testutils/test_utils.go [hunk 11]
```diff
@@ -328,14 +451,52 @@ func (cfg *IntegrationTest) TestIntegration(t *testing.T) {
...
+									cfg.ExecuteQuery(ctx, t, []string{currentTestTable}, "drop", false)
+									cfg.ExecuteQuery(ctx, t, []string{currentTestTable}, "create", false)
+									cfg.ExecuteQuery(ctx, t, []string{currentTestTable}, "add", false)
+									continue
+								}
+								// Delete old parquet files before schema evolution
+								if test.operation == "cleanup-parquet-files" {
+									if err := DeleteParquetFiles(t, cfg.DestinationDB, currentTestTable); err != nil {
+										return fmt.Errorf("failed to cleanup parquet files: %w", err)
+									}
```
Comment (vikaxsh) [inline utils/testutils/test_utils.go]: why are we deleting old parquet here ?

File: utils/testutils/test_utils.go [hunk 12]
```diff
@@ -460,15 +621,106 @@ func VerifyIcebergSync(t *testing.T, tableName, icebergDB string, datatypeSchema
...
+
+	parquetPath := fmt.Sprintf("s3a://warehouse/%s/%s", parquetDB, tableName)
+	viewName := fmt.Sprintf("`%s_view_%d`", tableName, time.Now().UnixNano())
+
+	// create a temporary view for parquet files, allows to run describe query
+	createViewQuery := fmt.Sprintf(
+		"CREATE OR REPLACE TEMP VIEW %s AS SELECT * FROM parquet.`%s`",
+		viewName, parquetPath,
+	)
+	_, err = spark.Sql(ctx, createViewQuery)
```
Comment (vikaxsh) [inline utils/testutils/test_utils.go:947]: we should add retry logic here as well similar to iceberg

File: utils/testutils/test_utils.go [hunk 13]
```diff
@@ -328,14 +451,52 @@ func (cfg *IntegrationTest) TestIntegration(t *testing.T) {
...
+										}
+									}
+								}
 								return nil
 							}
 
-							// 3. Run Sync command and verify records in Iceberg
+							// 3. Run Sync command and verify records in Iceberg and Parquet
 							for _, test := range testCases {
+								if test.operation == "reset" {
```
Comment (vikaxsh) [inline utils/testutils/test_utils.go]: we need to reset state file as well

File: utils/testutils/test_utils.go (hunk 12 above)
Comment (vaibhav-datazip) [reply utils/testutils/test_utils.go:947]: why would that be needed ?

Comment (vikaxsh) [reply utils/testutils/test_utils.go:947]: if spark failed to query for first time

Comment (vaibhav-datazip) [reply utils/testutils/test_utils.go:947]: I don't think spark will fail in this case, the issue earlier occurred in iceberg cases due to latency in update for iceberg catalog.

File: utils/testutils/test_utils.go (hunk 13 above)
Comment (vaibhav-datazip) [reply utils/testutils/test_utils.go]: we are running full refresh for next case , which will override the state file, so reset state wont be needed.

File: drivers/mongodb/internal/mon_test_util.go (hunk 9 above)
Comment (vaibhav-datazip) [reply drivers/mongodb/internal/mon_test_util.go:112]: evolve and update can occur simultaneously
